### PR TITLE
This PR fixes #5248 - parent host status is not persistent

### LINKED ISSUE #5248: parent host status is not persistent across restarts.
https://github.com/apache/trafficserver/blob/master/src/traffic_server/HostStatus.cc#L206

Should be changed to persistent.

### PR DESCRIPTION
**Pull request by jrushford**
This is a WIP PR to fix issue #5248.  Still testing.

Fixes #5248 across restarts.

### CODE REVIEW

File: src/traffic_server/HostStatus.cc [hunk 1]
```diff
@@ -94,6 +94,30 @@ mgmt_host_status_down_callback(ts::MemSpan span)
...
+
+  if (rec) {
+    // parse the hostname from the stat name
+
+    char *s = const_cast<char *>(rec->name);
+    // 1st move the pointer past the stat prefix.
+    s += strlen(stat_prefix.c_str());
+    hostname = s;
+    reason   = hostname.substr(hostname.find('_'));
+    reason.erase(0, 1);
```
Comment (vmamidi) [inline src/traffic_server/HostStatus.cc]: can you write a comment here?

File: src/traffic_server/HostStatus.cc [hunk 2]
```diff
@@ -94,6 +94,30 @@ mgmt_host_status_down_callback(ts::MemSpan span)
...
+    // erase the reason tag
+    hostname.erase(hostname.find('_'));
+    if (rec->data.rec_int == 0 && Reasons::validReason(reason.c_str())) {
```
Comment (vmamidi) [inline src/traffic_server/HostStatus.cc]: Can you specify that you are trying to look for hosts that were down. Also, How are we going to track time for hosts that were marked down for a certain period of time?

File: src/traffic_server/HostStatus.cc [hunk 3]
```diff
@@ -113,6 +137,14 @@ HostStatus::~HostStatus()
   ink_rwlock_destroy(&host_statids_rwlock);
 }
 
+void
+HostStatus::loadStats()
+{
+  if (RecLookupMatchingRecords(RECT_ALL, stat_prefix.c_str(), handle_record_read, nullptr) != REC_ERR_OKAY) {
+    Error("[HostStatus] - Error reading stats.");
```
Comment (vmamidi) [inline src/traffic_server/HostStatus.cc]: Maybe the error should say that error while readings for updating the host status?

File: src/traffic_server/traffic_server.cc [hunk 4]
```diff
@@ -1865,7 +1865,7 @@ main(int /* argc ATS_UNUSED */, const char **argv)
     initCacheControl();
     IpAllow::startup();
     ParentConfig::startup();
-    HostStatus::instance();
+    HostStatus::instance().loadStats();
```
Comment (vmamidi) [inline src/traffic_server/traffic_server.cc]: Should we call this function loadHostStatusFromStats() or something to indicate that this function is trying to load the Host Status from stats?

File: proxy/HostStatus.h [hunk 5]
```diff
@@ -86,6 +86,7 @@ struct HostStatus {
   void setHostStatus(const char *name, const HostStatus_t status, const unsigned int down_time, const char *reason);
   HostStatus_t getHostStatus(const char *name);
   void createHostStat(const char *name);
+  void loadStats();
```
Comment (vmamidi) [inline proxy/HostStatus.h]: Should we call this function loadHostStatusFromStats() or something to indicate that this function is trying to load the Host Status from stats?

**Comment by oknet** [comment]
The PR breaks SocksProxy.

The defaults socks server is set to UP (status: 2) during `loadSocksConfiguration()` which called by `netProcessor.init()`:
```
Traffic Server 9.0.0 Apr 29 2019 20:13:46 bogon
traffic_server: using root directory '/usr/local/ats'
[Apr 30 00:15:33.600]  DEBUG: <DNS.cc:1785 (ink_dns_init)> (dns) ink_dns_init: called with init_called = 0
[Apr 30 00:15:33.600]  DEBUG: <Socks.cc:474 (loadSocksConfiguration)> (Socks) Socks Version 5
[Apr 30 00:15:33.600]  DEBUG: <Socks.cc:484 (loadSocksConfiguration)> (Socks) server connect timeout: 10 socks respnonse timeout 15
[Apr 30 00:15:33.600]  DEBUG: <Socks.cc:495 (loadSocksConfiguration)> (SocksProxy) Read SocksProxy info: accept_enabled = 1 accept_port = 10800 http_port = 80
[Apr 30 00:15:33.600]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_manual, id: -1
[Apr 30 00:15:33.600]  DEBUG: <HostStatus.cc:249 (createHostStat)> (host_statuses) stat name: proxy.process.host_status.172.22.21.211_active, id: 1
[Apr 30 00:15:33.600]  DEBUG: <HostStatus.cc:249 (createHostStat)> (host_statuses) stat name: proxy.process.host_status.172.22.21.211_local, id: 2
[Apr 30 00:15:33.654]  DEBUG: <HostStatus.cc:249 (createHostStat)> (host_statuses) stat name: proxy.process.host_status.172.22.21.211_manual, id: 3
[Apr 30 00:15:33.654]  DEBUG: <HostStatus.cc:249 (createHostStat)> (host_statuses) stat name: proxy.process.host_status.172.22.21.211_self_detect, id: 4
[Apr 30 00:15:33.654]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_manual, id: 3
[Apr 30 00:15:33.654]  DEBUG: <HostStatus.cc:168 (setHostStatus)> (host_statuses) set status up for :  name: 172.22.21.211, status: 2, reason_stat: proxy.process.host_status.172.22.21.211_manual
[Apr 30 00:15:33.654]  DEBUG: <HostStatus.cc:177 (setHostStatus)> (host_statuses) name: 172.22.21.211, status: 2
[Apr 30 00:15:33.673]  DEBUG: <Socks.cc:502 (loadSocksConfiguration)> (Socks) Socks Config File: /usr/local/ats/etc/trafficserver/socks.config
[Apr 30 00:15:33.676]  DEBUG: <Socks.cc:530 (loadSocksConfiguration)> (Socks) Socks Turned on
```

The defaults socks server is set to DOWN (status: 1) by `HostStatus::instance().loadHostStatusFromStats()`.

```
[Apr 30 00:15:33.677]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_active, id: 1
[Apr 30 00:15:33.677]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_active, id: 1
[Apr 30 00:15:33.677]  DEBUG: <HostStatus.cc:172 (setHostStatus)> (host_statuses) set status down for :  name: 172.22.21.211, status: 1, reason_stat: proxy.process.host_status.172.22.21.211_active
[Apr 30 00:15:33.677]  DEBUG: <HostStatus.cc:177 (setHostStatus)> (host_statuses) name: 172.22.21.211, status: 1
[Apr 30 00:15:33.678]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_local, id: 2
[Apr 30 00:15:33.678]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_local, id: 2
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:172 (setHostStatus)> (host_statuses) set status down for :  name: 172.22.21.211, status: 1, reason_stat: proxy.process.host_status.172.22.21.211_local
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:177 (setHostStatus)> (host_statuses) name: 172.22.21.211, status: 1
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_manual, id: 3
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_manual, id: 3
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:172 (setHostStatus)> (host_statuses) set status down for :  name: 172.22.21.211, status: 1, reason_stat: proxy.process.host_status.172.22.21.211_manual
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:177 (setHostStatus)> (host_statuses) name: 172.22.21.211, status: 1
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_self_detect, id: 4
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:267 (getHostStatId)> (host_statuses) name: proxy.process.host_status.172.22.21.211_self_detect, id: 4
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:172 (setHostStatus)> (host_statuses) set status down for :  name: 172.22.21.211, status: 1, reason_stat: proxy.process.host_status.172.22.21.211_self_detect
[Apr 30 00:15:33.742]  DEBUG: <HostStatus.cc:177 (setHostStatus)> (host_statuses) name: 172.22.21.211, status: 1
```

```
1808   netProcessor.init();                     // ---> socks config loaded at here
1809   prep_HttpProxyServer();
...snip...
1862   } else {
1863     remapProcessor.start(num_remap_threads, stacksize);
1864     RecProcessStart();
1865     initCacheControl();
1866     IpAllow::startup();
1867     HostStatus::instance().loadHostStatusFromStats();    // ---> HostStatus persistent data loaded at here
1868     ParentConfig::startup();          // ---> parents proxy config loaded at here
1869 #ifdef SPLIT_DNS
1870     SplitDNSConfig::startup();
1871 #endif
```

@jrushford Does the HostStatus persistent data have to be loaded before the records ?

**Comment by zwoop** [comment]
This has merge conflicts, and we'll need a new 8.1.x PR. Also, feel free to recommend / suggest other PRs that this depends on, such that it makes the back porting easier.

**Pull request merged**